src/odemis/driver/pmtctrl.py

```diff
@@ -421,17 +423,41 @@ def _sendCommand(self, cmd):
         cmd (byte str): command to be sent to PMT Control unit.
         returns (byte str): answer received from the PMT Control unit
         raises:
-            IOError: if an ERROR is returned by the PMT Control firmware.
+            PMTControlError: if an ERROR is returned by the PMT Control firmware.
+            HwError: in case the of connection timeout
         """
         cmd = cmd + b"\n"
         with self._ser_access:
             logging.debug("Sending command %s", to_str_escape(cmd))
-            self._serial.write(cmd)
+            try:
+                self._serial.write(cmd)
+            except IOError:
+                logging.warn("Failed to send command to PMT Control firmware, "
+                             "trying to reconnect.")
+                if self._recovering:
+                    raise
+                else:
+                    self._tryRecover()
+                    # send command again
+                    logging.debug("Sending command %s again after auto-reconnect" % cmd)
+                    return self._sendCommand(cmd[:-1])  # cmd without \n
 
             ans = b''
             char = None
             while char != b'\n':
-                char = self._serial.read()
+                try:
+                    char = self._serial.read()
+                except IOError:
+                    logging.warn("Failed to read from PMT Control firmware, "
+                                 "trying to reconnect.")
+                    if self._recovering:
+                        raise
+                    else:
+                        self._tryRecover()
+                        # don't send command again
```

Comment:
What is the rationale behind not sending the command again?

Reply:
At this stage, we have already sent a command to the hardware and the connection failed afterwards. Sending the command a second time could potentially be harmful (like doing a move twice). In the PMT, there shouldn't be any commands that break something when sent twice, but it's better to be safe.